otwcode/otwarchive
# AO3-5905 New hit count code (for compatibility with nginx caching)

## Issue

https://otwarchive.atlassian.net/browse/AO3-5905

## Purpose

This PR rewrites the current hit count code so the hit count continues to work even when full page caching is enabled for logged-out users. The changes include:
- The work and chapter pages embed a short `<script>` element that posts to `HitCountController#create`, which calls `RedisHitCounter` to (potentially) increase the hit count.
- The hit count code in `RedisHitCounter` works as follows:
  - It computes the timestamp for the current 24 hour period, and combines that timestamp with the work ID to create a Redis key.
  - In an atomic operation, it adds the user's IP address to the set stored at that Redis key, and simultaneously adds the Redis key to a hash mapping all keys to the corresponding timestamp.
  - If the IP address is new (i.e. if `sadd` returns `1`/`true`), increment the Redis hash mapping from work IDs to recent hits counts. (Note that we can't combine this with the atomic operation for adding the IP address, because Redis "transactions" are basically just a list of commands to perform in sequence. As far as I can tell, there's no way to do conditional things like checking the results of calling `sadd`.)
- Every 30 minutes, the `RedisHitCounter` is scheduled to do the following:
  - Move the current hit count hash to a temporary key, so that it's not simultaneously being updated and modified.
  - Iterate through the hash, look up the `StatCounter` associated with each key, and increment that `StatCounter`'s `hit_count` by the value of the hit count hash.
  - Delete counts from the hash as soon as the corresponding changes to the StatCounter have been written to the database.
- Every 24 hours, the `RedisHitCounter` is scheduled to do the following:
  - Compute the current timestamp.
  - Iterate through the list associating keys (work ID/timestamp pairs) to their timestamp.
  - Delete any keys (and their corresponding sets) which have a timestamp less than the current timestamp.
- I included a migration to delete the `last_visitor` column from the `stat_counters` table. (I don't think it's currently used -- the last visitor info seems to be stored in Redis, currently -- but it's even more unnecessary now that the `last_visitor` isn't stored anywhere.)
- I included a migration to delete the `last_visitor_old` and `hit_count_old` columns from the `works` table. (I don't think either of them are currently used, but this seems as good a time as any to get rid of them.)
- I deleted a bunch of the old code for reading hit counts from logs and storing/retrieving hit counts from Redis, because the `RedisHitCounter` class controls all of that now. I also added a one-time task to delete all of the old keys from Redis.
- I modified the JS used to fetch `token_dispenser.json`, both to fix an apparent bug (setting `value` on the `<meta>` tag), and to make the code trigger a custom event when the token is updated (so that the query sent to the `HitCountController` includes the most up-to-date CSRF token).

I wasn't entirely sure how to check for cases where the hit count shouldn't be incremented: works in an unrevealed collection, works hidden by admin, drafts, etc. Obviously, that's not the sort of thing that we can check in the `HitCountController`, because we do *not* want to hit the database. So I ended up deciding to check it only when deciding whether to include the JS (on a page where the work and all of its attributes are already loaded).

I could have also included the check in `RedisHitCounter`, so that hits aren't added to works that shouldn't receive hits. But that could cause valid hits to be suppressed (if the work becomes inaccessible soon after some people have viewed it), or invalid hits to be counted (if the work becomes accessible soon after some admins/challenge moderators have viewed it).

## Testing Instructions

1. Make sure that the owner's hits don't count:
   - Find a work that you created, but haven't visited recently.
   - View it multiple times logged-in as the creator. Note the hit count in the header.
   - Wait at least 30 minutes for the stats to update.
   - View the work and verify that the hit count hasn't increased.
2. Make sure that logged-in hits count:
   - Find a work that you haven't visited recently.
   - Log in as a user unrelated to the work.
   - View the work multiple times. Note the hit count in the header.
   - Wait at least 30 minutes for the stats to update.
   - View the work and verify the hit count has increased by 1.
3. Make sure that logged-out hits count:
   - Find a work that you haven't visited recently.
   - View it multiple times logged-out. Note the hit count in the header.
   - Wait at least 30 minutes for the stats to update.
   - View the work and verify the hit count has increased by 1.
4. Make sure that cached hits count:
   - Find a work that you haven't visited recently.
   - Get two devices with different IP addresses ready (or one device with a proxy).
   - While logged-out, view the work on the first device (to generate the cache). Note the hit count in the header.
   - While logged-out, view the work on the second device.
   - Wait at least 30 minutes for the stats to update.
   - View the work again and verify that the hit count has increased by 2.
5. Make sure that interleaving multiple views only counts twice:
   - Find a work that you haven't visited recently.
   - Get two devices with different IP addresses ready (or one device with a proxy).
   - Make sure you're logged out (or logged in as a user other than the owner).
   - View the work on the first device.
   - View the work on the second device.
   - View the work on the first device again.
   - Wait at least 30 minutes for the stats to update.
   - View the work again and verify that the hit count has increased by 2.
6. Make sure that the rollover happens properly:
   - Find a work that you haven't visited recently.
   - Make sure you're logged out (or logged in as a user other than the owner).
   - View the work slightly before 3 AM UTC. Note the hit count in the header.
   - View the work again slightly after 3 AM UTC, with the same IP address.
   - Wait at least 30 minutes for the stats to update.
   - View the work again and verify that the hit count has increased by 2.
7. Make sure that viewing later chapters increments the hit count.
   - Find a multichapter work that you haven't visited recently.
   - Make sure you're logged out (or logged in as a user other than the owner).
   - Copy the work URL and add `/navigate` to the end of the path to view the chapter index.
   - Click on one of the later chapters. Note the hit count in the header.
   - Wait at least 30 minutes for the stats to update.
   - View the chapter again and verify that the hit count has increased by 1.

It would also be good to redo the tests from https://otwarchive.atlassian.net/browse/AO3-5639.
